src/types.ts

```diff
@@ -10,11 +10,18 @@ export type IMethodName =
   | 'EmailSubmission/changes'
   | 'EmailSubmission/set';
 
+export type IErrorName = 'error';
+
+export type IInvocationName = IMethodName | IErrorName;
+
 /**
  * See https://jmap.io/spec-core.html#the-invocation-data-type
- * [ name, arguments, id ]
  */
-export type IMethodCall = [IMethodName, IArguments, string];
+export type IInvocation<ArgumentsType> = [
+  name: IInvocationName,
+  arguments: ArgumentsType,
+  methodCallId: string,
```

Comment:
Can we have a IMethodCallId type to not mix it with other strings?

Reply:
I don't get it. Specs say `method call id` is an arbitrary string. It cannot be more specific. If I create a type, I go further from the specs.

Reply:
I think we don't need to wrap every primitive type property into separate type.

Reply:
I'm influenced by my scala background where I would just write:

```
case class MethodCallId(value: String) extends AnyVal
```

Getting type safety without runtime costs in one line of code, and being sure I could not pass a random string there without explicitly transforming it. 

Overall my experience is that strong typing allowed us to catch and solve many bugs on James side.

> I don't get it. Specs say method call id is an arbitrary string. It cannot be more specific. If I create a type, I go further from the specs.

Yes but that any string is a method call id. Maybe passing some other kind of strings (like mailbox id) there do not make sense... Strong typing prevents that.

Another benefit: as a new comer to the code base, having strings all over the place is hard: I do not know what they correspond to... While if they have a strong type, it is very explicit...

But I'm peeky and this work is already good as is IMO.

Reply:
When one would write code you would get notified, that you are passing some wrong thing into Foo, because it requires Bar type as and argument, or you return from Foo, wrong type, when it should return Baz.

But in this case whether one would pass `IMethodCallId` or `string`, wouldn't basically change anything, it would still be represented via string and not some complex type e.g. `IEmailBodyPart` containing bunch of different properties.

At least this my view of things.

Reply:
![image](https://user-images.githubusercontent.com/550970/117453467-462eab80-af45-11eb-8c77-3c59e3dcdf07.png)
We see that the third is method call id now we have named tuples :)